tensilelite/Tensile/Toolchain/Source.py

```diff
@@ -0,0 +1,217 @@
+# of this software and associated documentation files (the "Software"), to deal
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in
+# all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+#
+################################################################################
+
+import itertools
+import os
+import re
+import shlex
+import shutil
+import subprocess
+from pathlib import Path
+from typing import Iterable, List, Union
+
+from ..Common import globalParameters, print2,  ensurePath, ParallelMap2, splitArchs
+
+class SourceToolchain:
+    def __init__(self, compiler: str, bundler: str, buildIdKind: str, asanBuild: bool=False, saveTemps: bool=False):
+        self.compiler = compiler
+        self.bundler = bundler
+        self.buildIdKind = buildIdKind
+        self.asanBuild = asanBuild
+        self.saveTemps = saveTemps
+
+    def compile(self, srcPath: str, destPath: str, includePath: str, gfxs: List[str]):
+        """Compiles a source file into an object file.
+
+        Args:
+            cmdlineArchs: List of architectures for offloading.
+            kernelFile: The path to the kernel source file.
+            buildPath: The build directory path.
+            objectFilename: The name of the output object file.
+            outputPath: The output directory path.
+            globalParameters: A dictionary of global parameters.
+
+        Raises:
+            RuntimeError: If the compilation command fails.
+        """
+        launcher = shlex.split(os.environ.get("Tensile_CXX_COMPILER_LAUNCHER", ""))
+
+        hipFlags = [
+            "-D__HIP_HCC_COMPAT_MODE__=1",
+            "--cuda-device-only",
+            "-x", "hip", "-O3",    
+            "-I", includePath,
+            "-Xoffload-linker", f"--build-id={self.buildIdKind}",
+            "-std=c++17",
+        ]
+        if self.asanBuild:
+            hipFlags.extend(["-fsanitize=address", "-shared-libasan", "-fuse-ld=lld"])
+        if self.saveTemps:
+            hipFlags.append("--save-temps")
+        if os.name == "nt":
+            hipFlags.extend(["-fms-extensions", "-fms-compatibility", "-fPIC", "-Wno-deprecated-declarations"])
```

Comment:
Seems like we should know much of this stuff up front and can lift it out of this function.

Reply:
I'm not sure I understand what your vision of these functions are. Eventually, we must translate a boolean value to a string literal. If we lift them out of this function, where do we put them? Why not leave compiler-specific setup right where we create the compiler flag?

Reply:
Every time we call this function we are doing redundant work. I would like to only do it once.